dcpy/connectors/socrata/publish.py

```diff
@@ -0,0 +1,350 @@
+        rev = view.revisions.lookup(self.revision_num)
+        with open(path, "rb") as shp_zip:
+            logger.info(
+                f"Pushing shapefiles at {path} to {self.four_four} - rev: {self.revision_num}"
+            )
+            push_resp = (
+                rev.create_upload(dest_filename or shp_zip.name)
+                .shapefile(shp_zip)
+                .wait_for_finish()
+            )
+        error_details: dict | None = push_resp.attributes["failure_details"]
+        if error_details:
+            logger.error(f"Shapefile upload failed with {error_details}")
+            raise Exception()
+        return Socrata.Responses.RevisionDataSource(push_resp)
+
+    def upload_attachment(
+        self, path: Path, *, file_name: str | None = None
+    ) -> Socrata.Inputs.Attachment:
+        """Upload an attachment to a revision.
+
+        Note: uploading an attachment isn't enough to truly attach it to a revision.
+        It must also be patched into the metadata revision. Unfortunately, subsequent
+        patches seem to wipe out the attachment metadata, everything must be gathered
+        up into one patch.
+        """
+        # TODO: check other attachment types, e.g. xlsx
+        content_types = {".csv": "text/csv", ".pdf": "application/pdf"}
+
+        with open(path, "rb") as f:
+            attachment_md_raw = _socrata_request(
+                f"{self.revision_endpoint}/attachment",
+                "POST",
+                headers={
+                    "X-File-Name": file_name or path.name,
+                    "Content-type": content_types[path.suffix],
+                },
+                data=f,
+            )
+        return Socrata.Inputs.Attachment(
+            # TODO: diff between name and filename? and what about blob_id?
+            name=attachment_md_raw["filename"],
+            filename=attachment_md_raw["filename"],
+            asset_id=attachment_md_raw["file_id"],
+            blob_id=None,
+        )
+
+
+def push_shp_from_s3(dataset_name: str, version: str):
+    """Push Shapefile from edm-packaging to Socrata, and sync metadata."""
+    dataset_metadata: dict = datasets[dataset_name]
+    four_four: str = dataset_metadata["four_four"]
+    socrata_metadata = dataset_metadata["metadata"]
+    logger.info(
+        f"Syncing shapefile and metadata for {dataset_name}/{version} to Socrata {four_four}"
+    )
+
+    download_root_path = Path(".library")
+    dataset_path = download_root_path / dataset_name / version
+    logger.info(
+        f"Downloading shapefile from s3 dataset: {dataset_name} to {dataset_path.absolute()}"
+    )
+
+    s3.download_folder(
+        DISTRIBUTIONS_BUCKET, f"{dataset_name}/{version}/", download_root_path
+    )
+
+    dataset = Dataset(four_four=four_four)
+    dataset.discard_open_revisions()
+
+    rev = dataset.create_replace_revision()
+
+    data_source = rev.push_shp(path=dataset_path / "shapefile.zip")
```

Comment:
What happens if we successfully push a shapefile but we have a failure anywhere between here and applying the revision below? Does the shapefile get published or is kept as a part of the open revision?

Reply:
Good question - it's the latter. There'll be an open revision that folks can investigate, and close/apply manually